### PR TITLE
Refactors related to separating ast creating and linting

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add yourself to CONTRIBUTORS if you are a new contributor.
- [x] Write a good description on what the PR does.

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :hammer: Refactoring   |

## Description

Refs https://github.com/PyCQA/pylint/issues/7263.

### CODE REVIEW

File: pylint/lint/pylinter.py [hunk 1]
```diff
@@ -623,7 +626,10 @@ def check(self, files_or_modules: Sequence[str] | str) -> None:
 
         files_or_modules is either a string or list of strings presenting modules to check.
         """
+        # 1) Initialize
         self.initialize()
+
+        # 2) Gather all files
```
Comment (Pierre-Sassoulas) [inline pylint/lint/pylinter.py:632]: We could create a `_gather_all_files` fonction and remove comments.

File: pylint/lint/pylinter.py [hunk 2]
```diff
@@ -899,7 +912,7 @@ def _astroid_module_checker(
 
     def get_ast(
         self, filepath: str, modname: str, data: str | None = None
-    ) -> nodes.Module:
+    ) -> nodes.Module | None:
         """Return an ast(roid) representation of a module or a string.
 
         :param str filepath: path to checked file.
```
Comment (Pierre-Sassoulas) [inline pylint/lint/pylinter.py:918]:
```suggestion
        :param filepath: path to checked file.
```

File: pylint/lint/pylinter.py [hunk 3]
```diff
@@ -639,23 +645,30 @@ def check(self, files_or_modules: Sequence[str] | str) -> None:
                     "Missing filename required for --from-stdin"
                 )
 
-            filepath = files_or_modules[0]
+        # 3) Get all FileItems
```
Comment (Pierre-Sassoulas) [inline pylint/lint/pylinter.py:648]: Same with `_get_all_fileitems`.

**Review by Pierre-Sassoulas — COMMENTED** [review]
This would require another refactor but wouldn't it be a good thing to create à 'FileGatherer' class to have something testable and decoupled from the configuration in order to decrease the responsabilities of the PyLinter class ?

**Comment by DanielNoord** [comment]
> This would require another refactor but wouldn't it be a good thing to create à 'FileGatherer' class to have something testable and decoupled from the configuration in order to decrease the responsabilities of the PyLinter class ?

Wouldn't that be a bit overkill? I think it makes more sense to focus on creating a `FileChecker` which is a step somewhere in this refactoring process 😄

**Comment by Pierre-Sassoulas** [comment]
> I think it makes more sense to focus on creating a FileChecker which is a step somewhere in this refactoring process 

I'm not opinionated about creating a particular class or about a particular design, you know this part of the code better than me so you'll know better than me what to do :)

But I think we need to be able to test those changes at some point, right now we need to test through the PyLinter which means it's hard and we don't. The core internal of pylint (like pylint.message, and some part of the PyLinter) should be unitary tested (well, unitary testable is a first step) .

**Comment by DanielNoord** [comment]
> I'm not opinionated about creating a particular class or about a particular design, you know this part of the code better than me so you'll know better than me what to do :)

That's not necessarily true 😅 

> But I think we need to be able to test those changes at some point, right now we need to test through the PyLinter which means it's hard and we don't. The core internal of pylint (like pylint.message, and some part of the PyLinter) should be unitary tested (well, unitary testable is a first step) .

Yeah agreed. For now I'd say let's merge this and #7288. That would definelty set us up for the creation of `FileChecker` or something like that.

**Comment by Pierre-Sassoulas** [comment]
> Yeah agreed. For now I'd say let's merge this and https://github.com/PyCQA/pylint/pull/7288. That would definelty set us up for the creation of FileChecker or something like that.

👍

**Review by Pierre-Sassoulas — APPROVED** [review]

**Pull request merged**